modules/tinymce/src/themes/silver/main/ts/ui/context/ContextToolbarBounds.ts

```diff
@@ -6,13 +6,51 @@
  */
 
 import { Bounds, Boxes } from '@ephox/alloy';
-import { SelectorFind, SugarElement, WindowVisualViewport } from '@ephox/sugar';
+import { Optional } from '@ephox/katamari';
+import { Scroll, SelectorFind, SugarBody, SugarElement, SugarNode, Traverse, WindowVisualViewport } from '@ephox/sugar';
 
 import Editor from 'tinymce/core/api/Editor';
 
 import * as Settings from '../../api/Settings';
 import { UiFactoryBackstageShared } from '../../backstage/Backstage';
 
+const isVerticalOverlap = (a: Bounds, b: Bounds): boolean =>
+  Math.max(a.y, b.y) <= Math.min(a.bottom, b.bottom);
+
+const getRangeRect = (rng: Range): DOMRect => {
+  const rect = rng.getBoundingClientRect();
+  // Some ranges (eg <td><br></td>) will return a 0x0 rect, so we'll need to calculate it from the leaf instead
+  if (rect.height <= 0 && rect.width <= 0) {
+    const leaf = Traverse.leaf(SugarElement.fromDom(rng.startContainer), rng.startOffset).element;
+    const elm = SugarNode.isText(leaf) ? Traverse.parent(leaf) : Optional.some(leaf);
+    return elm.filter(SugarNode.isElement)
+      .map((e) => e.dom.getBoundingClientRect())
+      // We have nothing valid, so just fallback to the original rect
+      .getOr(rect);
+  } else {
+    return rect;
+  }
+};
+
+const getSelectionBounds = (editor: Editor): Bounds => {
```

Comment:
I wonder if this function may be useful elsewhere?

Reply:
At this stage no, as it's not that often you need to interact with the entire selection bounds. If we do have a case though, then we should definitely move it to somewhere more common.